src/ui/map.js

```diff
@@ -1045,18 +1056,51 @@ class Map extends Camera {
      */
     setProjection(projection?: ?ProjectionSpecification | string) {
         this._lazyInitEmptyStyle();
-        if (typeof projection === 'string') {
+        if (projection === undefined) { projection = null; } else if (typeof projection === 'string') {
             projection = (({name: projection}: any): ProjectionSpecification);
         }
-        this._runtimeProjection = projection;
-        this._transitionFromGlobe = false;
-        this.style.updateProjection();
-        return this;
+        this._updateProjection(projection);
     }
 
-    _setImplicitProjection(projection: ProjectionSpecification) {
-        this._runtimeProjection = projection;
-        this.style.updateProjection();
+    _updateProjection(projection?: ProjectionSpecification | null) {
+        const prevProjection = this.getProjection();
+        let changeExplicit = (projection !== undefined);
+        if (!changeExplicit) {
+            projection = prevProjection;
+        } else if (projection === null) {
+            this._explicitProjection = null;
+            changeExplicit = false;
+        }
+
+        // At high zoom when set to Globe, _runtimeProjection is Mercator while explicitProjection is globe.
+        if (projection && projection.name === "globe") {
+            projection = {name: (this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX ? 'mercator' : 'globe')};
+            if (changeExplicit) {
+                this._explicitProjection = {name: 'globe', center: [0, 0]};
+                changeExplicit = false;
+            }
+        }
+
+        const newProjection = this.transform.setProjection(projection);
+        if (!newProjection) return; // Continue if projection has changed
+
+        if (changeExplicit) {
+            this._explicitProjection = newProjection;
```

Comment:
I find the flow of this function quite difficult to follow, is there any way we could simplify it?
- Expose all early return conditions at the beginning of the function, otherwise wrap these entire blocks that need to be excluded in if statements, so that we don't have the [`return`](https://github.com/mapbox/mapbox-gl-js/pull/11460/files#diff-f62d8d42ec7424781f2e987a61684959e1bc794c67f75cd33548bd4a62c73028R1085) in the middle of the function
- Consolidate the cases for which we write the end result of `this._explicitProjection` and `changeExplicit` into a single write to make it easier to follow. We write to `this._explicitProjection` in 3 different locations in these statements.

Reply:
Pushed your first suggestion and I'm looking into the second.